pyzoo/zoo/ray/util/spark.py

```
@@ -0,0 +1,167 @@
+        self._pack_conda_main(["--output", tmp_path, "--n-threads", "8", "--name", conda_name])
+        print("Packing has been completed: {}".format(tmp_path))
+        return tmp_path
+
+    def _common_opt(self, master):
+        return '--master {} '.format(master)
+
+    def _create_sc(self, submit_args, conf):
+        from pyspark.sql import SparkSession
+        print("pyspark_submit_args is: {}".format(submit_args))
+        os.environ['PYSPARK_SUBMIT_ARGS'] = submit_args
+        zoo_conf = init_spark_conf()
+        for key, value in conf.items():
+            zoo_conf.set(key=key, value=value)
+        sc = init_nncontext(conf=zoo_conf, redirect_spark_log=self.redirect_spark_log)
+        sc.setLogLevel(self.spark_log_level)
+
+        return sc
+
+    def _init_yarn(self,
+                   python_zip_file,
+                   driver_memory,
+                   driver_cores,
+                   master=None,
+                   # settings for cluster mode
+                   executor_cores=None,
+                   executor_memory=None,
+                   extra_executor_memory_for_ray=None,
+                   #  settings for yarn only
+                   num_executor=None,
+                   spark_yarn_jars=None,
+                   penv_archive=None,
+                   hadoop_conf=None,
+                   hadoop_user_name=None,
+                   jars=None):
+        os.environ["HADOOP_CONF_DIR"] = hadoop_conf
+        os.environ['HADOOP_USER_NAME'] = hadoop_user_name
+        os.environ['PYSPARK_PYTHON'] = "python_env/bin/python"
+
+        def _yarn_opt(jars):
+            from zoo.util.engine import get_analytics_zoo_classpath
+            command = " --archives {}#python_env --num-executors {} " \
+                      " --executor-cores {} --executor-memory {}".\
+                format(penv_archive, num_executor, executor_cores, executor_memory)
+            path_to_zoo_jar = get_analytics_zoo_classpath()
```

this can possibly return BigDL class path, or return nothing at all?

---

yes, it would either return a zoo.jar or an empty string.

---

it can also possibly return bigdl.jar?

---

no, since it searches from the zoo distributed folder. Actually, I'm thinking we should throw an Exception here when the jar cannot be found then the user would aware that there's a problem here and can be fixed either by specify proper environment or reinstall Analytics-Zoo.

---

in `get_analytics_zoo_classpath`:

```
    if os.getenv("BIGDL_CLASSPATH"):
        return os.environ["BIGDL_CLASSPATH"]
```